File: DifferentiablePointRender/GaussianSplatting/src/ewa_project.cpp

#include <utils.h>
#include <ewa_project.h>
#include <ewa_project_kernel.h>


std::tuple<torch::Tensor, torch::Tensor, torch::Tensor>
EWAProjectForward(
    const torch::Tensor& xyz,
    const torch::Tensor& cov3d,
    const torch::Tensor& viewmat,
    const torch::Tensor& camparam,
    const torch::Tensor& uv,
    const torch::Tensor& depth,
    const torch::Tensor& xy,
    const int W, const int H,
    const torch::Tensor &visibility_status
){
    CHECK_INPUT(xyz);
    CHECK_INPUT(cov3d);
    CHECK_INPUT(viewmat);
    CHECK_INPUT(camparam);
    CHECK_INPUT(uv);
    CHECK_INPUT(depth);
    CHECK_INPUT(xy);
    CHECK_INPUT(visibility_status);

    const int P = xyz.size(0);
    auto float_opts = xyz.options().dtype(torch::kFloat32);
     auto int_opts = xyz.options().dtype(torch::kInt32);

    torch::Tensor cov2d = torch::zeros({P, 3}, float_opts);
    torch::Tensor radii = torch::zeros({P}, int_opts);
    torch::Tensor tiles_touched = torch::zeros({P}, int_opts);

    if(P != 0){
        EWAProjectForwardCUDA(
            P,
            xyz.contiguous().data_ptr<float>(),
            cov3d.contiguous().data_ptr<float>(),
            viewmat.contiguous().data_ptr<float>(),
            camparam.contiguous().data_ptr<float>(),
            uv.contiguous().data_ptr<float>(),
            depth.contiguous().data_ptr<float>(),
            xy.contiguous().data_ptr<float>(),
            W, H,
            visibility_status.contiguous().data_ptr<bool>(),
            cov2d.data_ptr<float>(),
            radii.data_ptr<int>(),
            tiles_touched.data_ptr<int>()
        );
    }

    return std::make_tuple(cov2d, radii, tiles_touched);
}

// std::tuple<torch::Tensor, torch::Tensor>
// EWAProjectBackward(
//     const torch::Tensor& xyz,
//     const torch::Tensor& cov3d,
//     const torch::Tensor& viewmat,
//     const torch::Tensor& projmat,
//     const torch::Tensor& camparam,
//     const int W, const int H,
//     const torch::Tensor& visibility_status,
//     const torch::Tensor& dL_dcov2d,
//     const torch::Tensor& dL_dopacity
// ){
//     CHECK_INPUT(xyz);
//     CHECK_INPUT(cov3d);
//     CHECK_INPUT(viewmat);
//     CHECK_INPUT(projmat);
//     CHECK_INPUT(camparam);
//     CHECK_INPUT(visibility_status);
//     CHECK_INPUT(dL_dcov2d);
//     CHECK_INPUT(dL_dopacity);

//     const int P = xyz.size(0);
//     auto float_opts = xyz.options().dtype(torch::kFloat32);

//     torch::Tensor dL_dxyz = torch::zeros({P, 3}, float_opts);
//     torch::Tensor dL_dcov3d = torch::zeros({P, 3}, float_opts);

//     if(P != 0){
//         // Kernel
//     }

//     return std::make_tuple(dL_dxyz, dL_dcov3d);
// }

File: DifferentiablePointRender/GaussianSplatting/ewa_project.py

import torch
from torch import Tensor
from typing import Tuple
from jaxtyping import Float, Bool
import DifferentiablePointRender.GaussianSplatting._C as _C

def ewa_project(
    xyz,
    cov3d,
    viewmat,
    camparam,
    uv,
    w, h,
    visibility_status=None
)->Tuple[Tensor, Tensor, Tensor]:
    
    if visibility_status is None:
        visibility_status = torch.ones_like(uv[:, 0], dtype=torch.bool)
        
    return _EWAProject.apply(
        xyz,
        cov3d,
        viewmat,
        camparam,
        uv,
        w, h,
        visibility_status
    )


class _EWAProject(torch.autograd.Function):
    @staticmethod
    def forward(
        ctx,
        xyz,
        cov3d,
        viewmat,
        camparam,
        uv,
        w, h,
        visibility_status
    ):
        
        (
            conic,
            radii,
            touched_tiles
        ) = _C.ewa_project_forward(
            xyz,
            cov3d,
            viewmat,
            camparam,
            uv,
            w, h,
            visibility_status
        )
        
        ctx.save_for_backward(
            xyz,
            cov3d,
            viewmat,
            camparam,
            radii
        )
        
        return conic, radii, touched_tiles
    
    
    @staticmethod
    def backward(
        ctx,
        dL_dconic,
        dL_dradii,
        dL_dtiles
    ):
        
        (
            xyz,
            cov3d,
            viewmat,
            camparam,
            radii
        ) = ctx.saved_tensors
        
        (
            dL_dxyz,
            dL_dcov3d
        ) = _C.ewa_project_backward(
            xyz,
            cov3d,
            viewmat,
            camparam,
            radii,
            dL_dconic
        )
        
        grads = (
            # loss gradient w.r.t xyz
            dL_dxyz,
            # loss gradient w.r.t cov3d
            dL_dcov3d,
            # loss gradient w.r.t viewmat
            None,
            # loss gradient w.r.t camparam
            None,
            # loss gradient w.r.t uv
            None,
            # loss gradient w.r.t W, 
            None,
            # loss gradient w.r.t H,
            None,
            # loss gradient w.r.t visibility_status
            None,
        )
        
        return grads